app/demo-loop/LDemoLauncher.hh

```diff
@@ -110,32 +113,94 @@ along_and_post_step_track(celeritas::CoreTrackView const& track)
     real_type step             = phys.step_length();
     real_type start_energy     = value_as<Energy>(particle.energy());
 
+    Action result_action = Action::unchanged;
+
     if (!particle.is_stopped())
     {
+        auto geo = track.make_geo_view();
+        auto mat = track.make_material_view();
+        auto rng = track.make_rng_engine();
+
+        // Apply the msc step limitation and pass the current step to geom
+        celeritas::detail::MscStepLimitResult msc_step_result;
+        if (phys.msc_ppid())
+        {
+            const celeritas::detail::UrbanMscNativeRef& urban_data
+                = phys.urban_data();
+
+            if (step > urban_data.params.geom_limit
+                && particle.energy() > urban_data.params.energy_limit)
+            {
+                using UrbanMscStepLimit = celeritas::detail::UrbanMscStepLimit;
+                UrbanMscStepLimit msc_step_limit(urban_data,
+                                                 particle,
+                                                 &geo,
+                                                 phys,
+                                                 mat.make_material_view(),
+                                                 sim);
+                msc_step_result = msc_step_limit(rng);
+
+                // Use msc_step_result.geom_path for geometry transport
+                step = msc_step_result.geom_path;
+
+                // Test for the msc limited step
+                if (msc_step_result.true_path < msc_step_result.phys_step)
+                {
+                    result_action = Action::msc_limited;
+                }
+            }
+        }
+
         // Boundary crossings and energy loss only need to be considered when
         // the particle is moving
         CELER_ASSERT(step > 0);
         {
```

Comment:
@whokion I was just looking over this again as part of the action refactor—if MSC is selected, shouldn't we skip the linear propagation? We already know that the distance is less than the safety so it won't hit the boundary, and we're already planning on changing the position and direction so we don't need to (or want to?) update the position along the old direction.

Reply:
Right. Thanks!